In [1]:
from tensorflow.keras import layers

In [2]:
from tensorflow import keras


In [3]:
inputs =keras.Input(shape=(28,28,1))
x= layers.Conv2D(filters=64,kernel_size=3,activation='relu')(inputs)
x= layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=128,kernel_size=3,activation='relu')(x)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=256,kernel_size=3,activation='relu')(x)
x = layers.GlobalAveragePooling2D()(x)
outputs = layers.Dense(10,activation='softmax')(x)
model = keras.Model(inputs= inputs, outputs = outputs)

In [4]:
model.summary(line_length=80)

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                      ┃ Output Shape             ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)          │ (None, 28, 28, 1)        │             0 │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ conv2d (Conv2D)                   │ (None, 26, 26, 64)       │           640 │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)      │ (None, 13, 13, 64)       │             0 │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)                 │ (None, 11, 11, 128)      │        73,856 │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)    │ (None, 5, 5, 128)        │             0 │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)                 │ (None, 3, 3, 256)        │       295,168 │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ global_average_pooling2d          │ (None, 256)              │             0 │
│ (GlobalAveragePooling2D)          │                          │               │
├───────────────────────────────────┼──────────────────────────┼───────────────┤
│ dense (Dense)                     │ (None, 10)               │         2,570 │
└───────────────────────────────────┴──────────────────────────┴───────────────┘

 Total params: 372,234 (1.42 MB)

 Trainable params: 372,234 (1.42 MB)

 Non-trainable params: 0 (0.00 B)


- 640 parameters kaise learn ya generate huvy?? -> kernal hai 3x3 yani 9 x feature map h 64 , to 9x64=576 , or har feature map k
- sth bais bhi hoga to +64

In [5]:
9*64+64

640

In [6]:
(9*128)*64+128 # ek imag se 64 feature map ay , jb isko input dia to 64 images(purane feature map) to har ek image se 128
                # feature map nikaly

73856

In [7]:
((9*256)*128)+256

295168

In [8]:
256*10+10

2570

--------------

In [9]:
from keras.datasets import mnist
(train_image,train_label),(test_image,test_label)=mnist.load_data()

In [10]:
train_image = train_image.reshape((60000,28,28,1))
train_image = train_image.astype('float32')/255
test_image = test_image.reshape((10000,28,28,1))
test_image = test_image.astype('float32')/255

In [ ]:
model.compile(
    optimizer="adam",
    loss="sparse_categorical_crossentropy",
    metrics=['accuracy'])

model.fit(train_image,train_label,epochs=4,batch_size=64)

Epoch 1/4
374/938 ━━━━━━━━━━━━━━━━━━━━ 2:46 296ms/step - accuracy: 0.7159 - loss: 0.8748